### ISSUE TITLE
[types] body.json() return type

### ISSUE THREAD

**Issue by SkeLLLa** [nodejs/undici#1562]
Currently `body.json()` returns `any` type. It's aligned to `JSON.parse` I guess. However from types perspective it's not very good.

For example without additional linters and configs we could do the following:
```typescript
const foo: any = 10; // We can assign anything to any
const bar: boolean = foo; // Any is assignable to anything 
foo.exec(); // Ok; anything goes with any
```
But if we have `unknown` type there it will be much safer and force end-users manually cast it to type that they expect.
```typescript
const foo: unknown = 10; // We can assign anything to unknown, but
const bar: boolean = foo; // Error, we can't assign unknown to any other type without an explicit assertion
foo.exec(); // Error, we can't access a property and call it on unknown
```

On the other hand it will be a breaking change from typescript point of view and maybe someone have other point of view regarding type safety in this case.

**Comment by meyfa** [comment]
In my opinion it makes sense to mirror types from `lib.dom.d.ts` so that code written for both the browser and for Node doesn't suffer from mismatched types. `lib.dom.d.ts` defines `body.json()` to be returning `any`, as well: https://github.com/microsoft/TypeScript/blob/298b3a432ca099330108ed7bb18293ae79693672/lib/lib.dom.d.ts#L2448

**Comment by mcollina** [comment]
I'm ok with `Promise<any>`

**Comment by Tksi** [comment]
It seems `body.json()` from `undici.request` returns `any` type, but `responce.json()` from `undici.fetch` returns `unknown` type.

```javascript
import { request, fetch } from 'undici';

request('http://httpbin.org/uuid')
  .then(({ body }) => body.json())
  .then((obj) => console.log(obj.uuid)); // OK, obj is type 'any'

fetch('http://httpbin.org/uuid')
  .then((responce) => responce.json())
  .then((obj) => console.log(obj.uuid)); // Type Error, obj is type 'unknown'
```

**Comment by KhafraDev** [comment]
> It seems body.json() from undici.request returns any type, but responce.json() from undici.fetch returns unknown type.

The first case was due to me trying to stay in line with the dom types and node-fetch's types (at that time; they've now changed to `unknown` as well).
https://github.com/microsoft/TypeScript/blob/298b3a432ca099330108ed7bb18293ae79693672/lib/lib.dom.d.ts#L2448
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/5b452194a34f86bb35532340ec19582a88337de2/types/node-fetch/index.d.ts#L149

Luckily, the body mixin types were duplicated instead of re-used, so that the fetch types always resolved to `unknown` and the non-fetch body mixin types resolved to `any`. I don't know if switching to unknown would be worth it, as it would cause issues in existing typescript projects.

**Comment by SkeLLLa** [comment]
Yes, changing the type will cause issues in existing projects, so if it's considered to be changed it should go in semver-major. But in general those issues will allow users to improve their projects.

For example, sometimes we use a customized ts compiler that treats all `any` as `unknown` to see weak points in code where `any` is used without casting to some type.

Here's also related issue in TS: https://github.com/microsoft/TypeScript/issues/26188

**Comment by imranbarbhuiya** [comment]
> Another good thing might be to adopt typescript generics. Cloudflare already does this in workers, and it works like a charm for us.
> 
> _aware of previous discussions, but even if Cloudflare adopts the pattern, I think for **interoperable runtimes** it would be great to allow this optionally_
> 
> ```ts
> fetch(…).then(res => res.json<JSONObject>())
> ```

https://github.com/DefinitelyTyped/DefinitelyTyped/pull/36868#issuecomment-512989650

**Comment by alexey-sh** [comment]
Another downside of using undici – its typescript support is poor.

**Comment by KhafraDev** [comment]
Thank you for the highly useful criticism. Everyone subscribed to this issue has now wasted time reading your comment.
